docs/design/debug-api.md

```diff
@@ -0,0 +1,352 @@
+endHash - string - the last hash of block at which to retreive data. Optional, defaults to startHash
+
+#### Returns 
+addresses - array of addresses
+
+#### Example Request
+
+```JSON
+{
+	"jsonrpc":"2.0",
+	"method":"debug_getModifiedAccountsByHash",
+	"params":[
+		"0x2a1af018e33bcbd5015c96a356117a5251fcccf94a9c7c8f0148e25fdee37aec",
+		"0x4e3d3e7eee350df0ee6e94a44471ee2d22cfb174db89bbf8e6c5f6aef7b360c5"
+	],
+	"id":"1"
+}
+```
+
+#### Example Response
+```JSON
+ {
+  "jsonrpc": "2.0",
+  "id": 1,
+    "result": [
+      "0x8c0041566e0bc27efe285a9e98d0b4217a46809c",
+      "0x3ee18b2214aff97000d974cf647e7c347e8fa585",
+      "0x76364611e457b1f97cd58ffc332ddc7561a193f6"
+    ]
+ }
+```
+
+### Method description
+
+`debug_getModifiedAccountsByNumber` - Returns all accounts that have changed between the two blocks specified.
+
+#### Parameters
+startNum - number - start block number
+endNum - number - end block number. Optional, defaults to startNum
+
+#### Returns 
+addresses - array of addresses
+
+#### Example Request
+```JSON
+{
+	"jsonrpc":"2.0",
+	"method":"debug_getModifiedAccountsByNumber",
+	"params":[
+		100,
+		200
+	],
+	"id":"1"
+}
+```
+
+#### Example Response
+```JSON
+ {
+  "jsonrpc": "2.0",
+  "id": 1,
+    "result": [
+      "0x8c0041566e0bc27efe285a9e98d0b4217a46809c",
+      "0x3ee18b2214aff97000d974cf647e7c347e8fa585",
+      "0x76364611e457b1f97cd58ffc332ddc7561a193f6"
+    ]
+ }
+```
+
+## Error Codes
+
+| Error Codes |        Error message        |                              Solution                               |
+| :---------: | :-------------------------: | :-----------------------------------------------------------------: |
+|    32000    |      Transaction not found.      | Occurs when user is trying to trace non-existing transaction. |
+|    32001    | Block hash not found. |          Occurs when user is trying to get modified accounts for non-existing block hash.           |
+|    32002    | Block number not found. |          Occurs when user is trying to get modified accounts for non-existing block number.           |
+
+## Limits
+1. Trying to get modified accounts for more than ex. 100 blocks. Env. variable can be `DEBUG_MODIFIED_ACCOUNTS_LIMIT`
+
+
+## Metric Capturing
+
+Capture metrics for the following:
+
+1. Log every call to all filter API method, as a total amount.
+2. Log every success or fail for each new API method, as a total amount.
+
+## Tests
+The following test cases should be covered but additional tests would be welcome.
+
+1. Overall functionality of the methods.
+2. Test different scenarios with all possible parameter combinations.
+3. Case where transaction is not found for `debug_traceTransaction`.
+4. Case where block hash is not found for `debug_getModifiedAccountsByHash`.
+5. Case where block number is not found for `debug_getModifiedAccountsByNumber`.
+
+
+## Deployment
+
+Debug API will run alongside the already available HTTP server.
```

Comment:
does this means, it will be part of the same "Relay" module right?
I recommend that is disabled by default and that these methods are available only if enabled.
`DEBUG_API_ENABLED=false`

Reply:
Yes, they are part of the relay module.
Once exposed I'm not sure that we can just disabled them, more like throw an error if they are disabled like -> Method Disabled when you try to call debug_traceTransaction for example.

Reply:
Thanks, that is what I thought with respect to the `alongside` but wanted to confirm.

This is for `TRACE` API, but is similar for `DEBUG`

```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "error": {
        "code": -32600,
        "message": "trace_filter is not available on the Free tier - upgrade to Growth, Scale, or Enterprise for access. See available methods at https://docs.alchemy.com/alchemy/documentation/apis"
    }
}
```

I am thinking that maybe `DEBUG` methods end up being too expensive to offer them for Free for Node Operators, and they will want to be able to enable them or disable them on certain deployments (PAID vs FREE)

Reply:
There should be a config for this correct. If disabled then we should return a not supported error like we do with other methods.
`debug_traceTransaction` should not be expensive for the callTrace case but will be for opcode flow.
callTrace should be enabled by default and opcode disabled by default per our plan.